DQM/Integration/python/clients/sistrip_dqm_sourceclient-live_cfg.py

```diff
@@ -243,7 +243,7 @@
     process.SiStripAnalyserCosmic.TkMapCreationFrequency  = -1
     process.SiStripAnalyserCosmic.ShiftReportFrequency = -1
     process.SiStripAnalyserCosmic.StaticUpdateFrequency = 5
-    process.SiStripAnalyserCosmic.MonitorSiStripBackPlaneCorrection = False
+    process.SiStripAnalyserCosmic.MonitorSiStripBackPlaneCorrection = cms.bool(False)
```

Comment:
why is this needed?

Reply:
It looks like it is indeed needed because that parameter is not set in SiStripAnalyserCosmic module but inherited from another config.
We (@pmandrik and I) found these errors when running DQM clients in python3 on 12_0_X for the very first time. It seems that dropping type specs only works if the parameters are explicitely set in the cloned modules, those inherited via C++ class inheritance or deference (like this) are not taken.
However, I am not a python expert, so any advice is welcomed, specially because this looks like a general problem, I wonder if nobody has faced it before us.

Reply:
@makortel

Reply:
I see the `process.SiStripAnalyserCosmic` apparently comes from
https://github.com/cms-sw/cmssw/blob/a19e4f667b438dc92d99069a3dfd5de526b52db3/DQM/SiStripMonitorClient/python/SiStripClientConfigP5_Cosmic_cff.py#L5
and that indeed does not have `MonitorSiStripBackPlaneCorrection` parameter defined. Therefore setting that paremeter here requires the explicit type. Perhaps the parameter should be defined in `DQM.SiStripMonitorClient.SiStripClientConfigP5_Cosmic_cff`? Then setting it here would not need an explicit type. (the cfi fragments should really contain the complete set of parameters).

I'm not really sure what exactly you meant with
> parameter ... inherited from another config

or
> those inherited via C++ class inheritance or deference (like this) are not taken.

-------

(as a side note, the `SiStripAnalyzer` appears to be a legacy module and needs to be migrated to one of the thread-friendly module types)

Reply:
> perhaps the parameter should be defined in DQM.SiStripMonitorClient.SiStripClientConfigP5_Cosmic_cff?

indeed, but I was wondering why it worked in py2. Or did it? @jfernan2 

> (as a side note, the SiStripAnalyzer appears to be a legacy module and needs to be migrated to one of the thread-friendly module types)

That's on my todo list.

Reply:
> indeed, but I was wondering why it worked in py2. Or did it? @jfernan2
It did not work, well, we did not even try since these type spec removal was done at 12_0_X when it was master and we are actually testing it now at P5

Reply:
> I'm not really sure what exactly you meant with

I meant cases like this or this other example modified in this PR:
DQM/EcalCommon/python/CommonParams_cfi.py
DQM/Integration/python/clients/hcalcalib_dqm_sourceclient-live_cfg.py

where the parameter subsystem is defined indirectly though C++ class inheritance, and set in the corresponding config of the original mother class, but not in the son/grandson cfi

Reply:
Could you be more specific? The `DQM/EcalCommon/python/CommonParams_cfi.py` does not have anything that would relate to C++ class inheritance, and `DQM/Integration/python/clients/hcalcalib_dqm_sourceclient-live_cfg.py` has too many things to really get a grasp.

Regardless of the structure of the C++ code, it is the responsibility of the concrete EDModule to define its full cfi fragment, either via its `fillDescriptions()` function (preferred), or in a hand-crafted `cfi.py` file.

Reply:
Sure, I will focus on:
DQM/Integration/python/clients/hcalcalib_dqm_sourceclient-live_cfg.py
L186 and L195 modify subsytem parameter in module qie11Task_laser = process.qie11Task.clone()

qie11Task is defined in:
https://github.com/cms-sw/cmssw/blob/edb013d63e3b1a934c974d310587366e3f5b242c/DQM/HcalTasks/python/QIE11Task.py

which is an instance of C++ plugin QIE11Task 
https://github.com/cms-sw/cmssw/blob/edb013d63e3b1a934c974d310587366e3f5b242c/DQM/HcalTasks/plugins/QIE11Task.cc

but does not set subsystem parameter because this parameter comes from inheritance of DQTask class:
https://github.com/cms-sw/cmssw/blob/6d2f66057131baacc2fcbdd203588c41c885b42c/DQM/HcalCommon/src/DQTask.cc

which does NOT set subsystem parameter either because it inherits too from class DQModule:
https://github.com/cms-sw/cmssw/blob/6d2f66057131baacc2fcbdd203588c41c885b42c/DQM/HcalCommon/src/DQModule.cc#L11
where there is a default value but at C++ code.

So, I could go and create a cfi for DQModule and set default values which are at C++ code, and then start calling it from python config files here and there...but I do not know if this is recommended at all...

Any advice is welcomed...

Reply:
Thanks for the explanation.

In the end the parameters need to end up in `DQM/HcalTasks/python/QIE11Task.py`. This can be achieved in (at least) three ways:

1. Add the parameters used by `QIE11Task`, `DQTask`, and `DQModule` manually to `qie11Task`. Leads to lot's of repetition if there are many other classes inheriting from `DQTask` and `DQModule`.
2. Create (handcrafted) PSets for `DQTask` and `DQModule`, and propagate the parameters from there to `qie11Task`. Eliminates the repetition of parameters, but exposes the inheritance structure in the configuration, that likely wouldn't need to know it.
3. Add `fillDescriptions()` function to `QIE11Task`, and `fillPSetDescription()` functions to `DQTask` and `DQModule`, such that the deriving class' function calls the parent class' function to add the parent class' parameters. Eliminates repetition, and the code organization (class hierarchy) stays as an implementation detail of `QIE11Task`.

The last one (3) would be the preferred pattern.

Reply:
Thanks @makortel 
Could you point me to a good example?
I have found this twiki:
https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideConfigurationValidationAndHelp#Converting_an_Existing_CFI_file
and I am able to create fillDescriptions() function to QIE11Task but I am not sure if it should go to .h file (protected, public section) or to .cc
And in case of the parent classes, I am not sure how to create fillPSetDescription() functions.
Thanks

Reply:
I have pushed my naive attempt to fulfill the comments, any advice is welcomed!
Thanks